src/os/bluestore/BlueStore.cc

```diff
@@ -15268,48 +15274,53 @@ int BlueStore::_deferred_replay()
       }
     );
   }
-  CollectionRef ch = _get_collection(coll_t::meta());
-  bool fake_ch = false;
-  if (!ch) {
-    // hmm, replaying initial mkfs?
-    ch = static_cast<Collection*>(create_new_collection(coll_t::meta()).get());
-    fake_ch = true;
-  }
-  OpSequencer *osr = static_cast<OpSequencer*>(ch->osr.get());
+  if (tracepoint_debug_deferred_replay_start) tracepoint_debug_deferred_replay_start();
+  IOContext ioctx(cct, nullptr);
   KeyValueDB::Iterator it = db->get_iterator(PREFIX_DEFERRED);
-  for (it->lower_bound(string()); it->valid(); it->next(), ++count) {
+  for (it->lower_bound(string()); it->valid(); /*iterator update outside*/) {
     dout(20) << __func__ << " replay " << pretty_binary_string(it->key())
 	     << dendl;
+    if (keys_to_remove) {
+      keys_to_remove->push_back(it->key());
+    }
     bluestore_deferred_transaction_t *deferred_txn =
       new bluestore_deferred_transaction_t;
     bufferlist bl = it->value();
     auto p = bl.cbegin();
     try {
       decode(*deferred_txn, p);
+
+      bool has_some = _eliminate_outdated_deferred(deferred_txn, bluefs_extents);
+      if (has_some) {
+        if (tracepoint_debug_deferred_replay_track) tracepoint_debug_deferred_replay_track(*deferred_txn);
+        for (auto& op: deferred_txn->ops) {
+          for (auto& e : op.extents) {
+            bufferlist t;
+            op.data.splice(0, e.length, &t);
+            bdev->aio_write(e.offset, t, &ioctx, false);
+          }
+        }
+      } else {
+        delete deferred_txn;
```

Comment:
Looks like deferred_txn is leaking if not getting here.
Previously TransContext cared about that...

Reply:
In fact we don't need it to be allocated from heap any more.

Reply:
No longer allocated.